### DISCUSSION TITLE
PHP 8.0 | Squiz.WhiteSpace.MemberVarSpacing: add support for attributes

### DISCUSSION THREAD

**Discussion by mandarinian** [squizlabs/PHP_CodeSniffer#3419] [Show and tell]
Hi, on 6 May I implemented support for attributes for class member vars. When I got feedback from @jrfnl, ideas were good and I've also implemented them. After that PR got stuck and on 29 Jul @gsherwood added commit 64f2a81e24517da68c4cdf6312983fc6be31037e that completely conflicting with my solution, but didn't solve important issues like:
1. Support for member vars with attributes, but without doc block.
2. Valid attribute positioning, including:
    * single attribute on line check
    * disallow blank lines between attributes, phpdocs and member vars
    * attributes must be after phpdoc
    * attributes can't be on the same line with members vars
    * valid positioning of multiline attributes

Also I implemented auto-fixes for all possible issues related to my PR.

My team using CS on a daily basis and we really miss a lot features implemented in my PR.
Is it possible to pay attention to this PR? #3334
Suggestions for fixes or improvements are welcome.
Thank you in advance :raised_hands:

**Comment by jrfnl** [comment]
@yariknechyporuk I can't speak for the project or the maintainer(s), but if you want my opinion:

Your PR was created without an issue attached to it and without a description explaining what you are changing and why, which generally results in a PR which is low on the priority list.

Your PR also changes an existing sniff which is checking for something specific: number of lines of whitespace between properties.

Your change is attempting to make that sniff do something new and may well break existing rulesets which expect the sniff to only target whitespace between properties. This makes your PR problematic.

What I believe is needed is:
* A sniff specifically for attributes which ruleset maintainers can then choose to add to their rulesets or not, but which is not included in any PHPCS native rulesets by default, i.e. sniffs in the `Generic` standard.
* This sniff could then examine *all* types of attributes, be it class, property, function etc.
* I can even imagine several sniffs in an `Attributes` category. Thinking out loud:
    - A sniff to verify the formatting *within* attributes (whitespace on the inside of the brackets, indentation in multi-line attributes, whether blank lines within an attribute are allowed etc)
    - A sniff to disallow multi-line attributes.
    - A sniff to enforce one attribute per line.
    - A sniff to verify the spacing *around* the attribute in the wider context of the code, i.e enforce blank line above if there is no docblock, enforce no blank line after, maybe indentation of the attribute opener (if not already handled by the `ScopeIndent` sniff) etc
    - A sniff to enforce that attributes are always **above** a potential docblock belonging to the same structure.
    - A sniff to enforce that attributes are always **between** a potential docblock and the structure.

For lack of a PSR covering the formatting of attributes, we need to give ruleset maintainers options and make this elective.

If at a later point in time PSR would come up with rules covering attributes, we can then select which of the above sniffs cover those rules and add those to a ruleset to cover the (new) PSR rules.

Hope this helps.
